common/src/main/java/io/netty/util/concurrent/AutoScalingEventExecutorChooserFactory.java

```diff
@@ -0,0 +1,338 @@
+            this.executors = executors;
+            this.allExecutorsChooser = DefaultEventExecutorChooserFactory.INSTANCE.newChooser(executors);
+            this.activeChildrenCount = new AtomicInteger(maxChildren);
+
+            this.activeExecutors = executors;
+            this.activeExecutorsChooser = DefaultEventExecutorChooserFactory.INSTANCE.newChooser(executors);
+
+            ScheduledFuture<?> utilizationMonitoringTask = GlobalEventExecutor.INSTANCE.scheduleAtFixedRate(
+                    new UtilizationMonitor(), utilizationCheckPeriodNanos, utilizationCheckPeriodNanos,
+                    TimeUnit.NANOSECONDS);
+
+            if (executors.length > 0) {
+                executors[0].terminationFuture().addListener(future -> utilizationMonitoringTask.cancel(false));
+            }
+        }
+
+        /**
+         * This method is only responsible for picking from the active executors list.
+         * The monitor handles all scaling decisions.
+         */
+        @Override
+        public EventExecutor next() {
+            // Read the volatile fields once to get a consistent snapshot.
+            EventExecutor[] active = this.activeExecutors;
+            EventExecutorChooser chooser = this.activeExecutorsChooser;
+
+            if (active.length == 0) {
+                // This is only reachable if minChildren is 0 and the monitor has just suspended the last active thread.
+                // To prevent an error and ensure the group can recover, we wake one up and use the
+                // chooser that contains all executors as a safe temporary choice.
+                tryScaleUpBy(1, true);
+                return allExecutorsChooser.next();
+            }
+            return chooser.next();
+        }
+
+        /**
+         * Helper method to rebuild the active executors list and chooser.
+         */
+        private void updateActiveExecutors() {
+            List<EventExecutor> active = new ArrayList<>(activeChildrenCount.get());
+            for (EventExecutor executor : executors) {
+                if (!executor.isSuspended()) {
+                    active.add(executor);
+                }
+            }
+            EventExecutor[] newActiveExecutors = active.toArray(new EventExecutor[0]);
+            this.activeExecutors = newActiveExecutors;
+            this.activeExecutorsChooser = DefaultEventExecutorChooserFactory.INSTANCE.newChooser(newActiveExecutors);
+        }
+
+        /**
+         * Tries to increase the active thread count by a given amount, respecting the maxChildren limit.
+         * This method is thread-safe.
+         *
+         * @param amount    The desired number of threads to add to the active count.
+         * @param tryWakeUp If true, this method will also attempt to wake up a suspended thread.
+         * @return The number of threads that were actually woken up (will be 0 if tryWakeUp is false).
+         */
+        private int tryScaleUpBy(int amount, boolean tryWakeUp) {
```

Comment:
`tryWakeUp` is always `true` and this method doesn't really do anything if it's `false` anyway… we can remove it.